bus: remove process comptime parameter

matklad opened the pull request:
We include both replica and client busses in the binary (for the recover
command), so its better to cut down code bloat here.

matklad committed:
bus: remove process comptime parameter

We include both replica and client busses in the binary (for the recover
command), so its better to cut down code bloat here.

matklad committed:
bus: separate message bus initialization from opening the socket.

The direct motivation here is that, without this change, tb_context.zig
fails to compile. The reason fro that is that the return type of `init`
(`!void`) differs between clients and replicas. Clients don't initialize
the listening  socket, and don't have associated errors.

Now that we have erased the type level distinction between the two,
tb_context.zig has to deal with this error types, even if they don't
arise dynamically.

I considered two ways to fix it:

* split `init_client` and `init_replica` methods
* split `listen` as separate from `init`

Both feel fine, and there's an argument that maybe we actually need to
implement both regardless, but given that _today_ signatures of inits
are the same, I decided to punt on that part for now.

And I like a separate listen, as that's a very distinct IO operation,
which is visible by the outside world, so its helpful to be explicit
about when we are ready to talk to someone!

Long term, I actually want to pull _initialization_ of message bus up to
main, and pass in the  Bus (rather than BusOptions) down to replica &
client, but that's for some future work!

matklad committed:
bus: adapt to the new API

matklad committed:
bus: simplify fuzzer

No need to care about NodeAny anymore!

I am on the fence where we should do node.message_bus.deinit or
node.deinit, but I went for the simplest solution (not introducing
deinit).

matklad commented:
```
error: lld-link: undefined symbol: AcceptEx
    note: referenced by D:\a\tigerbeetle\tigerbeetle\src\io\windows.zig:377
    note:               tb_client.lib(tb_client.lib.obj):(io.windows.IO.accept__anon_39308__struct_41077.do_operation__anon_41973)
```

This one is going to be fun!

EDIT: surprisingly, it was a fun-fun and not a dreadful fun, af54374. I guess, thanks, Microsoft, for baroque .dll structure which revealed an unoptimality in the untyped message bus, which is now fixed!

matklad committed:
bus: don't codegen accept path into clients

What a puzzling diff it is! Here's a clue:

    error: lld-link: undefined symbol: AcceptEx
        note: referenced by D:\a\tigerbeetle\tigerbeetle\src\io\windows.zig:377
        note:               tb_client.lib(tb_client.lib.obj):(io.windows.IO.accept__anon_39308__struct_41077.do_operation__anon_41973)

So what happens here is that most of WinSock functionality we use is
available in ws2_32.dll, which we link to. However, on accept path we
use `ws2_32.AcceptEx` function which, despite its Zig name, actually
comes from mswsock.dll. A direct fix would be to add dependency on that
.dll to our clients, but I think we have a different problem here.

The clients _shouldn't_ have accept path at all! While it'll never be
called, just having the code generated there in the binary is
suboptimal! Luckily, we can rely on Zig's lazy compilation semantics to
not generate calls to accept, if it is statically obvious that accept is
not called. Which we can do by separating ticks for client and the
replica!

I don't wan't to be super pedantic here, and I want to allow replicas
and clients to share the same tick, if they end up being in the same
process, so I don't forbid calling the general tick from the client, and
just add a specialization (calling specialized tick from replica would
be a mistake, and that is asserted).

matklad committed:
bus: tighten the contract

On the second thought, I don't think there's a reason to not be strict
here, and require replicas to tick, and clients to tick_client. This is
easy to assert, so we shouldn't be able to mess it up!

Best defensive programming is offensive programming!

sentientwaffle commented on src/message_bus.zig:
```diff
@@ -389,13 +384,13 @@ pub fn MessageBusType(comptime IO: type, comptime process_type: vsr.ProcessType)
         /// Failure is silent and returns the connection to an unused state.
         fn connect_connection(bus: *MessageBus, connection: *Connection, replica: u8) void {
-            if (process_type == .replica) assert(replica != bus.process.replica);
+            if (bus.process == .replica) assert(replica > bus.process.replica);
```
:heart:

sentientwaffle commented on src/message_bus.zig:
```diff
@@ -389,13 +384,13 @@ pub fn MessageBusType(comptime IO: type, comptime process_type: vsr.ProcessType)
             const family = bus.configuration[replica].any.family;
-            connection.fd = init_tcp(bus.io, family) catch |err| {
+            connection.fd = init_tcp(bus.io, bus.process, family) catch |err| {
                 log.err("{}: connect_to_replcia: init_tcp error={s}", .{
```
s/replcia/replica

sentientwaffle commented on src/message_bus.zig:
```diff
@@ -1020,20 +1017,18 @@ pub fn MessageBusType(comptime IO: type, comptime process_type: vsr.ProcessType)
+                .client, .client_likely => |client_id| {
+                    // A newer client connection may have replaced this one:
+                    if (bus.clients.get(client_id)) |existing_connection| {
+                        assert(bus.process == .replica);
```
Can this assert move up out of the branch, since clients don't talk to other clients?

sentientwaffle commented on src/message_bus_fuzz.zig:
```diff
@@ -17,22 +17,14 @@ const log = std.log.scoped(.message_bus_fuzz);
-const NodeReplica = NodeType(MessageBusReplica);
-const NodeClient = NodeType(MessageBusClient);
-
-/// TODO If we remove type-level distinction between MessageBusReplica and MessageBusClient, then we
-/// can simplify erase_types()/Context handling in the fuzzers IO implementation to avoid some
-/// `*any opaque`s. We can also remove Node's helper methods in favor of accessing the message_bus
-/// directly.
```
I think this TODO is only half-implemented (unless that is intentional?). The other half is to simplify the `*anyopaque`s/`erase_types` -- since the only consumer of this IO implementation is here, we can hard-code the context type to a `*MessageBus`.

matklad replied:
Ah, I wanted to add a comment here and forgot! 

Yeah, this is intentional! We _can_ hard-code `Context` as `*MessageBus`, but we'd still have to pass in a dummy `Context: type` argument, because that what's the bus does, and because IO interface _is_ generic in this way. 

So I figured that maybe we don't actually want to fix this part, and to keep close to the actual IO interface?

sentientwaffle replied:
I was thinking we could leave the interface the same but internally e.g.

```
    pub const Completion = struct {
        context: *MessageBus,
        callback: union(enum) {
            accept: // actual callback type, not type-erased
        },
        operation: Operation,
    };
```

Not sure if it would be more verbose after all, but the impetus is that I spent a day or so while writing the fuzzer fighting a weird bug which ended up being due to a bad `anyopaque` cast.

matklad replied:
Ah, I see now, yeah, like this very much, implemented in [5787766](https://github.com/tigerbeetle/tigerbeetle/pull/3370/commits/5787766e74013f7da94a46f881de5b3ca2752fd4)

matklad committed:
bus: code review

matklad committed:
bus: erase erase_types from the fuzzer

The casting is tricky here, and easy to get wrong, so let's strongly
type this!
